File: rummikub_solver/backend/src/models/turn.py
"""Turn ORM model."""

import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Boolean, DateTime, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import JSONB
from src.core.database import Base

class Turn(Base):
    __tablename__ = "turns"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    game_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("games.id"), index=True, nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String(500), nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="create", nullable=False)
    board_image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    hand_image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    results: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    is_first_drop: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

File: rummikub_solver/frontend/lib/features/games/domain/entities/turn.dart
// Turn metadata and lifecycle status.

enum TurnStatus { create }

class Turn {
  final String id;
  final String gameId;
  final String name;
  final String? description;
  final DateTime createdAt;
  final TurnStatus status;
  final bool isFirstDrop;

  const Turn({
    required this.id,
    required this.gameId,
    required this.name,
    this.description,
    required this.createdAt,
    required this.status,
    this.isFirstDrop = false,
  });

  factory Turn.fromJson(Map<String, dynamic> json) {
    return Turn(
      id: json['id'] as String,
      gameId: json['gameId'] as String,
      name: json['name'] as String? ?? 'Unnamed Turn',
      description: json['description'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : (json['timestamp'] != null
              ? DateTime.parse(json['timestamp'] as String)
              : DateTime.now()),
      status: TurnStatus.values.firstWhere(
        (status) => status.name == (json['status'] as String).toLowerCase(),
        orElse: () => TurnStatus.create,
      ),
      isFirstDrop: json['isFirstDrop'] as bool? ?? false,
    );
  }
}
